src/librustc_coverage/coverage.rs

```diff
@@ -0,0 +1,335 @@
+                        | ExprKind::Yield(result_expr) => {
+                            match result_expr.take() {
+                                Some(wrapped_expr) => {
+                                    *result_expr = Some(self.wrap_and_count_expr(
+                                        &block.span,
+                                        wrapped_expr,
+                                        report,
+                                    ));
+                                }
+                                None => {
+                                    block.stmts.push(self.count_stmt(
+                                        &block.span,
+                                        last.span.shrink_to_lo(),
+                                        report,
+                                    ));
+                                }
+                            }
+                            block.stmts.push(last);
+                        }
+                        ExprKind::Continue(..) => {
+                            block.stmts.push(self.count_stmt(
+                                &block.span,
+                                last.span.shrink_to_lo(),
+                                report,
+                            ));
+                            block.stmts.push(last);
+                        }
+                        _ => {
+                            let insert_after_last = last.span.shrink_to_hi();
+                            block.stmts.push(last);
+                            block.stmts.push(self.count_stmt(
+                                &block.span,
+                                insert_after_last,
+                                report,
+                            ));
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+    }
+}
+
+impl MutVisitor for CoverageVisitor<'_, '_> {
+    fn visit_block(&mut self, block: &mut P<Block>) {
+        self.instrument_block(block.deref_mut());
+        mut_visit::noop_visit_block(block, self);
+    }
+
+    fn flat_map_item(&mut self, item: P<Item>) -> SmallVec<[P<Item>; 1]> {
+        if let ItemKind::Fn(_defaultness, _signature, _generics, block) = &item.kind {
+            if item.ident.name == sym::main {
+                if let Some(block) = block {
+                    self.main_block_id = Some(block.id);
+                }
+            }
+            self.function_stack.push(item.ident.name);
+            let result = mut_visit::noop_flat_map_item(item, self);
+            self.function_stack.pop();
+            result
+        } else {
+            mut_visit::noop_flat_map_item(item, self)
+        }
+    }
+
+    // FIXME(richkadel):
+    // add more visit_???() functions for language constructs that are branched statements without
+    // blocks, such as:
+    //    visit match arm expr
+    // if not block, wrap expr in block and inject counter
```

Comment:
I'm not sure we care much about blocks in particular at all (unless they happen to be the block which defines a function body). Looking directly for branch expressions seems like what we want.

In clang focusing on blocks makes a bit more sense, since blocks usually correspond to some kind of branch in C++. (Even then it doesn't seem like exactly the right thing to me, but maybe I'm missing something.)

Reply:
that's a good point and simplifies the solution by not making non-block branch expressions a special case. Cool! 

I still need to support function blocks though, I believe.

Reply:
On second thought, I need to give that more thought ;-)

return (and yield), break, and continue may complicate the idea that this is simpler. I'll think about it.

Reply:
Just following up on this: I've been experimenting with various ways to inject the counters, in a number of different (and some obscure) branch types, and as it turns out (as best I can tell), my initial approach seems to be the best way to handle block expressions.

Non-block expressions are straightforward enough, but I can't apply the same simplicity to blocks because of the possibility that the block might contain a return, break, or continue.

(If the block doesn't contain a return, break, or continue, I think the block can be handled just like a non-block expression, but intuitively I think it's better to handle all blocks the same and not worry about what might or might not be inside them.)

For example:

```rust
match var1 {
    Pattern1 => expr(),
...
```

I can count the execution of the Pattern1 arm with:

```rust
match var1 {
    Pattern1 => count(..., expr()),
...
```
But (if I interpretted your suggestion correctly), to ignore blocks and just inject counters based on expressions, I tried the following (which certainly looks tempting):

```rust
match var1 {
    Pattern1 => count(..., {
        do_one_thing();
        get_result()
    }),
...
```

No problem there, but:

```rust
match var1 {
    Pattern1 => count(..., {
        println!("Bad pattern");
        return Err(...);
    }),
...
```

This will not compile because it is clear (to the compiler) that `count()` will never be executed.

So this is what led me to the current prototype solution for blocks, in this case:

```rust
match var1 {
    Pattern1 => {
        println!("Bad pattern");
        return count(..., Err(...));
    },
...
```

There are a few other edge cases like this that require handling blocks differently from non-block branched expressions. I'll try to cover these in future documentation, with my branch analysis.

Let me know if I missed something about your suggestion.

Thanks!